runtime/skiplist.c

```diff
@@ -0,0 +1,215 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Cambium, INRIA Paris                  */
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+/* A dictionary data structure implemented as skip lists
+   (see William Pugh, "Skip lists: a probabilistic alternative to
+   balanced binary trees", Comm. ACM 33(6), 1990). */
+
+#include <stddef.h>
+#include "caml/config.h"
+#include "caml/memory.h"
+#include "caml/misc.h"
+#include "caml/skiplist.h"
+
+/* Generate a random level for a new node: 0 with probability 3/4,
+   1 with probability 3/16, 2 with probability 3/64, etc.
+   We use a simple linear congruential PRNG (see Knuth vol 2) instead
+   of random(), because we need exactly 32 bits of pseudo-random data
+   (i.e. 2 * (NUM_LEVELS - 1)).  Moreover, the congruential PRNG
+   is faster and guaranteed to be deterministic (to reproduce bugs). */
+
+static uint32_t random_seed = 0;
+
+static int random_level(void)
+{
+  uint32_t r;
+  int level = 0;
+
+  /* Linear congruence with modulus = 2^32, multiplier = 69069
+     (Knuth vol 2 p. 106, line 15 of table 1), additive = 25173. */
+  r = random_seed = random_seed * 69069 + 25173;
+  /* Knuth (vol 2 p. 13) shows that the least significant bits are
+     "less random" than the most significant bits with a modulus of 2^m,
+     so consume most significant bits first */
+  while ((r & 0xC0000000U) == 0xC0000000U) { level++; r = r << 2; }
+  CAMLassert(level < NUM_LEVELS);
+  return level;
+}
+
+/* Initialize a skip list */
+
+void caml_skiplist_init(struct skiplist * sk)
+{
+  sk->forward[0] = NULL;
+  sk->level = 0;
+}
+
+/* Search a skip list */
+
+int caml_skiplist_find(struct skiplist * sk, uintnat key, uintnat * data)
+{
+  int i;
+  struct skipcell * e, * f;
+
```

Comment:
```suggestion

 CAMLassert(0 <= sk->level && sk->level < NUM_LEVELS);
 
```
(or a more general `SKIP_LIST_ASSERT_VALID` or something, given the number of places it might want to go)

Reply:
I went back and forth several times on this assertion business.  You're right that if we want well-formedness assertions we should have them in other functions too.

On the other hand, the invariant on `sk->level` is weak: it's relatively obvious from the code that it holds, while there are many other, more subtle ways for the skiplist structure to get corrupted.  

On the third hand, a full structural check is too costly as a runtime assertion.

In the end, I removed the two `CAMLassert` over `sk->level`.

Reply:
Do you already have code for the full structural check? (Only) If so, might that be worth including (surrounded by `#ifdef DEBUG`)? That would allow the possibility, for example, of relevant parts of the GC checking the structure just before running (i.e. check before sustained use, rather than check on every operation) similar to some of the heap validation stuff in the debug runtime?

Reply:
I have handwritten incomplete pseudocode for the full structural check...  I promise I will add it the first time we need to debug a corrupted skiplist.

Reply:
I'll hold you to being able to find that yellowed piece of paper :wink: